RCN example

shrinuKushagra opened the pull request:
This PR contains an example implementation of RCN using the pgmax package. 
We load a pre-trained RCN model on a very small subset of mnist (20 examples) and test on a small subset of mnist (20 examples).
The reported accuracy = 0.80.

StannisZhou commented on examples/rcn/load_data.py:
```diff
@@ -0,0 +1,78 @@
+# export
+import os
+
+import cv2
```
I don't think it's worth it to introduce an additional dependency just for resizing....You can use https://jax.readthedocs.io/en/latest/jax.image.html instead

StannisZhou commented on examples/rcn/load_data.py:
```diff
@@ -0,0 +1,78 @@
+
+
+# export
+def get_mnist_data_iters(data_dir, train_size, test_size, full_test_set=False, seed=5):
```
As discussed, replace this function with something that can download the MNIST datasets. It seems https://pypi.org/project/tensorflow-datasets/ is a standalone package that works well with JAX, so I suggest we add this to our dependency (by updating https://github.com/vicariousinc/PGMax/blob/master/pyproject.toml) and use it to access MNIST

StannisZhou commented on examples/rcn/load_data.py:
```diff
@@ -0,0 +1,78 @@
+    """
+    Load MNIST data.
+
+    Assumed data directory structure:
```
Not sure we want to assume this kind of data directory structure....Better to work with a loaded dataset where things are already organized into arrays.

StannisZhou commented on examples/rcn/load_data.py:
```diff
@@ -0,0 +1,78 @@
+            for fname in samples:
+                filepath = os.path.join(cat_path, fname)
+                # Resize and pad the images to (200, 200)
+                image_arr = cv2.resize(
```
Replace with `jax.image.resize` to minimize dependency

StannisZhou commented on examples/rcn/load_data.py:
```diff
@@ -0,0 +1,78 @@
+# export
```
If we use `tensorflow_datasets` to load MNIST digits, I see no reason for this to be a separate file. We should be able to do data loading and preprocessing in a few lines in the **STANDALONE** notebook

StannisZhou commented on examples/rcn/mnist_small.py:
```diff
@@ -0,0 +1,310 @@
+import numpy as np
+from jax import numpy as jnp
+from jax import tree_util
+from load_data import get_mnist_data_iters
```
Get rid of this module and make this notebook standalone

StannisZhou commented on pgmax/fg/graph.py:
```diff
@@ -48,6 +48,7 @@ class FactorGraph:
         Mapping[Hashable, groups.VariableGroup],
         Sequence[groups.VariableGroup],
         groups.VariableGroup,
+        Mapping[int, groups.VariableGroup],
```
This is very hacky....Could you investigate why this is happening and get a principled solution? I just tried and
```
from typing import Hashable
isinstance(2, Hashable)
```
gives me `True`. Also tests are still failing so this is not fixing the issue.

shrinuKushagra replied:
There does not seem to be a principled solution to this problem. Atleast, I could not find one. So have included the hack for now. 
python/mypy#2412

StannisZhou commented on examples/rcn/preproc.py:
```diff
@@ -0,0 +1,171 @@
+            filter_scale=filter_scale, num_orients=num_orients
+        )
+
+    def fwd_infer(self, img, brightness_diff_threshold=40.0):
```
Would be best to extract the relevant bits for this function and get rid of the redundant bits, and also move this into the standalone notebook. Directly copying this bit can be confusing for a tutorial-type example (especially when certain bits seem redundant).

StannisZhou commented on examples/rcn/mnist_small.py:
```diff
@@ -0,0 +1,310 @@
+hps, vps = 12, 12
+train_size = 20
+test_size = 20
+data_dir = "/storage/users/skushagra/MNIST/"
```
This is supposed to be included as a tutorial example....People won't have access to `/storage`

StannisZhou commented on examples/rcn/mnist_small.py:
```diff
@@ -0,0 +1,310 @@
+#
+
+# %%
+directory = f"/storage/users/skushagra/pgmax_rcn_artifacts/model_science_{train_size}_{hps}_{vps}"
```
Please include necessary data files inside `examples/example_data`, and structure the notebook in such a way that it can be run anywhere.

StannisZhou commented on examples/rcn/mnist_small.py:
```diff
@@ -0,0 +1,310 @@
+                unary_msg[v, index] = 1
+
+        evidence_updates[idx] = unary_msg
+    return bu_msg, evidence_updates
```
This `bu_msg` is no longer used anywhere. Could you please get rid of this?

StannisZhou reviewed (changes requested):
Several major issues:

1. We should aim for **A SINGLE NOTEBOOK FILE** (with a few data files in `examples_data`)
2. There seems to be many redundant/random pieces
3. Since this is intended as a tutorial, please make sure this can actually be run by someone. That means:
    - Make sure the data is accessible
    - Make sure you have properly updated the relevant dependicies
    - Make sure you have detailed docstrings/type hints for all your functions, and relevant comments when needed

shrinuKushagra commented:
Addressed the comments.

However, I could not find a principled solution to the `Hashable int` problem. 
Seems like there might be some issues with the library itself.
https://github.com/python/mypy/issues/2412

StannisZhou commented on examples/rcn.py:
```diff
@@ -0,0 +1,392 @@
+    """
+
+    mnist_train_size = 60000
+    dataset = fetch_openml("mnist_784", as_frame=False, cache=True)
```
Would be great if there is a way to get a progress bar here. If that's not possible, maybe also print another message reading `Fetching the data` before this line? Otherwise it's just silent and feels strange....

shrinuKushagra replied:
Added.

StannisZhou commented on examples/rcn.py:
```diff
@@ -0,0 +1,392 @@
+    data["filters"],
+)
+
+M = (2 * hps + 1) * (2 * vps + 1) + 1
```
Can we get rid of this `+ 1` as previously discussed? The 0 state does not make sense in this context and does not match the `science_rcn` implementation.

shrinuKushagra replied:
Changed.

StannisZhou commented on examples/rcn.py:
```diff
@@ -0,0 +1,392 @@
+    plt.text((c1 + c2) // 2, (r1 + r2) // 2, str(w), color="blue")
+    plt.plot([c1, c2], [r1, r2], color="blue", linewidth=0.5)
+
+plt.imshow(img, cmap="gray")
```
![Screenshot from 2021-11-20 16-11-50](https://user-images.githubusercontent.com/10180197/142744593-5d06f8a1-a268-4794-99ef-9aefa069eed7.png)
This visualization doesn't look good. Better to get rid of the boundary part, change coloring (I have lots of trouble reading the figure) by just using white background and use something other than blue for the perturb radiuses?

StannisZhou replied:
Thanks! Looking much better. Can we further zoom in to the actual model and get rid of all the blank space around it (so `img[60:130, 60:160]` or something similar you see fit)

shrinuKushagra replied:
Changed.

StannisZhou commented on pgmax/fg/graph.py:
```diff
@@ -45,7 +45,8 @@ class FactorGraph:
 
     variables: Union[
-        Mapping[Hashable, groups.VariableGroup],
+        Mapping[int, groups.VariableGroup],
```
Better to relax it to `Mapping[Any, groups.VariableGroup]` than restricting it to `int` and `str`, since there is no reason the names should be just `int` or `str`. I will also try to look into this

shrinuKushagra replied:
Changed.

StannisZhou reviewed (commented):
Made some initially comments after trying to run it. Will make more later

StannisZhou commented on examples/rcn.py:
```diff
@@ -0,0 +1,392 @@
+
+
+max_perturb_radii = 25
+phis = []
```
Change `phis` to more informative names

StannisZhou replied:
Reminder

StannisZhou commented on examples/rcn.py:
```diff
@@ -0,0 +1,392 @@
+        fg.add_factor(
+            [(idx, i1), (idx, i2)],
+            phis[r],
+            np.zeros(phis[r].shape[0]),
```
This can be omitted (it would be automatically initialized to all 0 if omitted)

StannisZhou replied:
Reminder

StannisZhou commented on examples/rcn.py:
```diff
@@ -0,0 +1,392 @@
+    num_orients = 16
+    brightness_diff_threshold = 40.0
+
+    filtered = np.zeros((filters.shape[0],) + img.shape, dtype=np.float32)
```
Add a visualization of the filters

StannisZhou commented on examples/rcn.py:
```diff
@@ -0,0 +1,392 @@
+                pos_chan[pos_chan > 0] *= factor
+            np.maximum(pc, pos_chan, pc)
+    bu_msg = np.array(pooled_channels)
+    return bu_msg
```
Add a visualization of the `bu_msg` we get from running preproc

StannisZhou commented on examples/rcn.py:
```diff
@@ -0,0 +1,392 @@
+    for idx in range(frcs.shape[0]):
+        frc = frcs[idx]
+
+        unary_msg = -1 + np.zeros((frc.shape[0], M))
```
Make sure your `M=625`

StannisZhou commented on examples/rcn.py:
```diff
@@ -0,0 +1,392 @@
+            evidence = bu_msg[f, r - hps : r + hps + 1, c - hps : c + hps + 1]
+            indices = np.transpose(np.nonzero(evidence > 0))
+
+            for index in indices:
```
This for loop looks quite inefficient. How is this different from directly doing `unary_msgs[v] = evidence.ravel()`?

shrinuKushagra replied:
Added.

StannisZhou commented on examples/rcn.py:
```diff
@@ -0,0 +1,392 @@
+        frc = frcs[idx]
+
+        unary_msg = -1 + np.zeros((frc.shape[0], M))
+        for v in range(frc.shape[0]):
```
A jax implementation for the whole thing would be
```
@functools.partial(jax.vmap, in_axes=(None, 0), out_axes=0)
def get_evidence(bu_msg, frc):
    ```
    bu_msg: Array of shape (n_features, M, N)
    frc: Array of shape (n_frcs, 3)
    ```
    return jax.lax.dynamic_slice(
        bu_msg[frc[0]], 
        jnp.array([frc[1] - hps, frc[2] - vps]), 
        jnp.array([2 * hps + 1, 2 * vps + 1])
    ).ravel()
```
Since it's vmapping you can directly apply this to the whole `frcs` array so also no more outer for loops.

shrinuKushagra replied:
I tried this implementation. Currently, have it commented out.
But when I ran this code, it was slower than the for loop (0.05 sec vs 0.4 sec). Maybe it will be faster when using larger size datasets?

StannisZhou replied:
Might be because no `jit` is used (I didn't include it in the above snippet). Let me test it and see. 

After the change to `ravel` having this for loop in the notebook is not too bad. I'll report back once I play with it a bit.

StannisZhou replied:
After adding `jit` timing becomes similar (still slightly slower at 0.08). But it seems to create some issues regarding memory usage so I still went with the CPU version.

StannisZhou reviewed (commented):
Made more comments

StannisZhou commented on examples/rcn.py:
```diff
@@ -0,0 +1,392 @@
+        unary_msg = -1 + np.zeros((frc.shape[0], M))
+        for v in range(frc.shape[0]):
+            f, r, c = frc[v, :]
+            evidence = bu_msg[f, r - hps : r + hps + 1, c - hps : c + hps + 1]
```
You are using `hps` twice here. It looks like a bug. A harmless one, but still a bug.

StannisZhou commented on examples/rcn.py:
```diff
@@ -0,0 +1,392 @@
+# # 1. Load the data
+# %%
+hps, vps = 12, 12
+train_size = 20
```
In my test, on a GPU with 10 GB of memory, with https://github.com/vicariousinc/PGMax/pull/97 you are now able to run inference with 100 templates in one batch. I suggest we change `train_size` to 100 (so that the detections/performances would look better).

shrinuKushagra replied:
Using `train_size = 100` has the following advantages. 
1. Backtraces might look better. 
2. Higher accuracy.

Disadvantages.
1. A person who does not have a powerful gpu will probably not be able to use it. 
2. Notebook takes a long time to finish on the CPU. For small train set, notebook finishes in a reasonable time even on the CPU. 

If the backtraces look decent on for the smaller case, might be better to use size = 20. Let me know what you think.

StannisZhou replied:
Good point on the requirement of a powerful GPU. Maybe we can give user both options? So load all 100 models, and subset to use only 20 if needed (a simple indexing change), and add a comment detailing the considerations/trade-off.

shrinuKushagra replied:
Added support for 100 templates but the default is still 20.
Can switch and wrote comments for explanation.

shrinuKushagra commented on examples/rcn.py:
```diff
@@ -0,0 +1,392 @@
+
+
+# %%
+train_set, train_labels, test_set, test_labels = fetch_mnist_dataset(
```
Removed `train_set`.

shrinuKushagra commented on examples/rcn.py:
```diff
@@ -0,0 +1,392 @@
+    train_set, train_labels = _sample_data(
+        full_train_set, full_train_labels, train_size // 10
+    )
+    test_set, test_labels = _sample_data(
```
Removed this abstraction.

shrinuKushagra commented on examples/rcn.py:
```diff
@@ -0,0 +1,392 @@
+        full_test_set, full_test_labels, test_size // 10
+    )
+
+    return train_set, np.array(train_labels), test_set, np.array(test_labels)
```
Removed.

shrinuKushagra commented on examples/rcn.py:
```diff
@@ -0,0 +1,392 @@
+
+
+# %% [markdown]
+# ## 6.1 Backtrace of some models on this test image
```
I changed the visualization to have backtraces of top model for some test images.
![0](https://user-images.githubusercontent.com/2974401/143890565-fac3cf33-3f76-4250-892f-e282f02bc130.png)
Overlaying seemed a bit non-intuitive. Let me know how this looks.

StannisZhou replied:
![Screenshot from 2021-11-29 20-01-45](https://user-images.githubusercontent.com/10180197/143983661-19ef548f-28af-4229-be26-ccb0b7831da7.png)
Since we do overlaying all the time not sure how it's non-intuitive. Do you mean the detection doesn't look that good? That is understandable as we are working with a small number of templates.

shrinuKushagra replied:
Added overlay.

StannisZhou commented on examples/rcn.py:
```diff
@@ -0,0 +1,392 @@
+    cols = []
+    index = 0
+    for i in range(M):
+        r1, c1 = -hps + i // (2 * vps + 1), -vps + i % (2 * vps + 1)
```
Reminder that this is still an outstanding comment (and should be an easy one to address). 

Even if the for loop is not that slow, this kind of direct manipulation would be hard to read (especially when there are readily available builtin functions for the same purpose).

For the `min/max` operations you can use array indexing to pick the relevant entries later.

shrinuKushagra commented:
All the comments are now addressed. 
Let me know once you run it and other comments you have.

StannisZhou commented:
Want to mention that not all comments are addressed. There were 3 unaddressed comments (with reminders), eg. https://github.com/vicariousinc/PGMax/pull/96#discussion_r754666201

But I addressed them in my commits. Left a few more comments from playing with the notebook. Once these are addressed we can merge the PR.

StannisZhou commented on examples/rcn.py:
```diff
@@ -0,0 +1,390 @@
+
+# %%
+test_set, test_labels = fetch_mnist_dataset(test_size)
+train_labels = (
```
I tried subsetting a set of trained models, and evaluation breaks completely. In general this seems a very unreliable way to get training labels. Please update `example_data/rcn_100.npz` to include:
1. The 100 training images
2. The labels for the 100 training images
3. The 100 trained RCN models

And load the `train_labels` from `rcn_100.npz`.

shrinuKushagra replied:
Updated. The file is now called `rcn.npz` and train sizes below 100 can be sub-sampled.

StannisZhou commented on examples/rcn.py:
```diff
@@ -0,0 +1,390 @@
+# %%
+data = np.load("example_data/rcn_100.npz", allow_pickle=True, encoding="latin1")
+frcs, edges, suppression_masks, filters = (
+    data["frcs"][:train_size],
```
Please include additional data so that this kind of subsetting works properly. Right now it seems it would break if I use values other than 20 or 100.

StannisZhou replied:
I now see that we need training images from each of the categories so simple sunsetting won't work....But it would be good in any case to have just one set of models and subset from that. One possibility is to change train size to something like  training image per class and so sunsetting accordingly.

shrinuKushagra replied:
Please refer to the reply above.

StannisZhou commented on examples/rcn.py:
```diff
@@ -0,0 +1,390 @@
+M = (2 * hps + 1) * (2 * vps + 1)
+
+# %% [markdown]
+# # 3. Visualize loaded model
```
For visualizing the model, please also visualize the corresponding training image side by side

StannisZhou commented on examples/rcn.py:
```diff
@@ -0,0 +1,390 @@
+    num_per_class = test_size // 10
+
+    print("Fetching the MNIST dataset")
+    dataset = fetch_openml("mnist_784", as_frame=False, cache=True)
```
From playing with the notebook, the caching does not seem to work. Please investigate and we should make caching work.

shrinuKushagra replied:
This seems to be a known issue.
https://github.com/scikit-learn/scikit-learn/issues/18783

Fixed using the suggestions in this thread.

StannisZhou commented on examples/rcn.py:
```diff
@@ -0,0 +1,390 @@
+# ## 5.2 Run map product inference on all test images
+
+# %%
+def get_evidence(bu_msg: np.ndarray, frc: np.ndarray):
```
Return type (I forgot to add)

shrinuKushagra replied:
Added.

StannisZhou reviewed (commented):
Hopefully the final round of comments

swaroopgj commented on examples/rcn.py:
```diff
@@ -0,0 +1,392 @@
+frc, edge = frcs[4], edges[4]
+fig, ax = plt.subplots(1, 1, figsize=(10, 10))
+for e in edge:
+    i1, i2, w = e
```
It might be worth adding some context here in comments or notebook cells that explain what the ws are.

shrinuKushagra replied:
Added.

wlehrach commented on examples/rcn.py:
```diff
@@ -0,0 +1,390 @@
+
+# %%
+def get_evidence(bu_msg: np.ndarray, frc: np.ndarray):
+    """Function to get evidence
+
+    Args:
+        bu_msg: Array of shape (n_features, 200, 200). Contains BU messages
+        frc: Array of shape (n_frcs, 3).
+
+    Returns:
+        evidence; Array of shape (n_frcs, M). Contains evidence
+    """
```
```suggestion
    """Returns the evidence (shape (n_frcs, M))

    Args:
        bu_msg: Array of shape (n_features, 200, 200). Contains BU messages
        frc: Array of shape (n_frcs, 3).
    """
```
You can make this concise if you want

> Returns: (or Yields: for generators)
>    Describe the type and semantics of the return value. If the function only returns None, this section is not required. It may also be omitted if the docstring starts with Returns or Yields (e.g. """Returns row from Bigtable as a tuple of strings.""") and the opening sentence is sufficient to describe the return value. 

https://google.github.io/styleguide/pyguide.html

shrinuKushagra replied:
Changed to this style.

swaroopgj commented on examples/rcn.py:
```diff
@@ -0,0 +1,392 @@
+    data["filters"],
+)
+
+M = (2 * hps + 1) * (2 * vps + 1)
```
Needs some explanation, unless the variable is named in a interpretable way.
For this particular variable, better to move where it is first used for better context.

shrinuKushagra replied:
Added.

swaroopgj commented on examples/rcn.py:
```diff
@@ -0,0 +1,392 @@
+# %%
+start = time.time()
+assert frcs.shape[0] == edges.shape[0]
+
```
I would assign `M` here so that it is clear what it is and why it is defined as such (different categorical values PGM variables can take and how they correspond to pool choices)

shrinuKushagra replied:
Added.

swaroopgj commented on examples/rcn.py:
```diff
@@ -0,0 +1,392 @@
+# ### 4.2.1 Pre-compute the valid configs for different perturb radii.
+
+# %%
+def valid_configs(r: int) -> np.ndarray:
```
Can you include pool sizes as inputs to this function? 
That makes it clear what is needed to compute these configs.

shrinuKushagra replied:
Added.

swaroopgj commented on examples/rcn.py:
```diff
@@ -0,0 +1,392 @@
+    return np.concatenate(configs)
+
+
+max_perturb_radius = 25
```
I would group this and hps,vps and any other model related params at one place and add a comment stating that.

shrinuKushagra replied:
Added configs at one place.

swaroopgj commented on examples/rcn.py:
```diff
@@ -0,0 +1,392 @@
+fg = graph.FactorGraph(variables=variables_all_models)
+for idx in range(edges.shape[0]):
+    edge = edges[idx]
+
```
Worth adding a comment on what you are adding here like how you are mapping RCN model edges to factor graph.

shrinuKushagra replied:
Added comment.

swaroopgj commented on examples/rcn.py:
```diff
@@ -0,0 +1,392 @@
+        img: The rgb image to compute bottom up messages on (H x W x 3).
+
+    Returns:
+        An array of shape [16 x H x W] denoting the presence or absence of an oriented line segment at a particular location.
```
These are not just oriented but also directional

shrinuKushagra replied:
Changed the text.

swaroopgj commented on examples/rcn.py:
```diff
@@ -0,0 +1,392 @@
+            The elements of this array belong to the set {+1, -1}.
+    """
+
+    num_orients = 16
+    brightness_diff_threshold = 40.0
```
Again, these can be grouped and defined along with model params if neededm, but these are not arbitrary as they should match the model you loaded, so read it from the loaded model (filters). BDT can be but not num_orients

shrinuKushagra replied:
Added configs at one place as suggested above.

swaroopgj commented on examples/rcn.py:
```diff
@@ -0,0 +1,392 @@
+        kern = filters[i, :, :]
+        filtered[i] = fftconvolve(img, kern, mode="same")
+
+    localized = np.zeros_like(filtered)
+    cross_orient_max = filtered.max(0)
+    filtered[filtered < 0] = 0
+    for i, (layer, suppress_mask) in enumerate(zip(filtered, suppression_masks)):
+        competitor_maxs = maximum_filter(layer, footprint=suppress_mask, mode="nearest")
+        localized[i] = competitor_maxs <= layer
+    localized[cross_orient_max > filtered] = 0
```
Add comment explaining what we are doing in this step

shrinuKushagra replied:
Added explanation.

swaroopgj commented on examples/rcn.py:
```diff
@@ -0,0 +1,392 @@
+    localized *= (filtered / brightness_diff_threshold).clip(0, 1)
+    localized[localized < 1] = 0
+
+    pooled_channel_weights = [(0, 1), (-1, 1), (1, 1)]
+    pooled_channels = [-np.ones_like(sf) for sf in localized]
+    for i, pc in enumerate(pooled_channels):
+        for channel_offset, factor in pooled_channel_weights:
+            ch = (i + channel_offset) % num_orients
+            pos_chan = localized[ch]
+            if factor != 1:
+                pos_chan[pos_chan > 0] *= factor
+            np.maximum(pc, pos_chan, pc)
```
Add comment explaining what we are doing in this step

shrinuKushagra replied:
Added explanation.

swaroopgj commented on examples/rcn.py:
```diff
@@ -0,0 +1,392 @@
+            np.maximum(pc, pos_chan, pc)
+
+    bu_msg = np.array(pooled_channels)
+    bu_msg[bu_msg == 0] = -1
```
Add comment explaining what you are doing here and why.

shrinuKushagra replied:
Added explanation.

swaroopgj commented on examples/rcn.py:
```diff
@@ -0,0 +1,392 @@
+        frc: Array of shape (n_frcs, 3).
+
+    Returns:
+        evidence; Array of shape (n_frcs, M). Contains evidence
```
```suggestion
        evidence: Array of shape (n_frcs, M). Contains evidence
```

shrinuKushagra replied:
Removed.

swaroopgj commented on examples/rcn.py:
```diff
@@ -0,0 +1,392 @@
+
+
+# %% [markdown]
+# ## 5.2 Run map product inference on all test images
```
map?

shrinuKushagra replied:
Expanded on this.

swaroopgj commented on examples/rcn.py:
```diff
@@ -0,0 +1,392 @@
+# # 6. Compute metrics (accuracy)
+
+# %%
+best_model_idx = np.argmax(scores, axis=1)
```
what if we have two classes that score the same?

shrinuKushagra replied:
`argmax` will chose the one with smaller value.

swaroopgj replied:
I know but that is wrongly reducing the accuracy.

shrinuKushagra replied:
If two templates have the same score, I am not sure what we can do in this case. It amounts to an arbitrarily selecting between the two (or more) of them.

swaroopgj reviewed (commented):
Thanks for adding. I have left comments regarding readability since these are intended for understanding.
If possible I would add a short text at the top to describe RCN model with reference to the paper and the Science implementation available.

shrinuKushagra commented:
> Want to mention that not all comments are addressed. There were 3 unaddressed comments (with reminders), eg. [#96 (comment)](https://github.com/vicariousinc/PGMax/pull/96#discussion_r754666201)
> 
> But I addressed them in my commits. Left a few more comments from playing with the notebook. Once these are addressed we can merge the PR.

Hi Stannis,
Sorry about this. I went back and checked which of the comments were not addressed. It seems that the following comments were missed.

1. Can we instead use np.unravel_index/ravel_multi_index for these?   -- In the context of the `valid_configs` function.
2. Changing the for loop in valid_config construction 
3 This can be omitted (it would be automatically initialized to all 0 if omitted) -- In the context of `valid_configs`

I went over the `Conversations` tab multiple times and these comments were not there. But looking back at the github emails, I do see that you had made these comments and also reminders for this. The only possible explanation seems that, these comments were present in the `Files` tab and not in the conversations tab. I just realized that sometimes there are comments that only show in the `Files` tab and not in `Conversations` tab. (this is in the context of another pr).

To conclude, I want to apologize for missing them. It was genuine over-sight (from not knowing about a github feature) and I will be more careful in future. Also, thanks for addressing the comments.

shrinuKushagra commented on examples/rcn.py:
```diff
@@ -0,0 +1,390 @@
+def fetch_mnist_dataset(test_size: int, seed: int = 5):
+    """Returns test images sampled randomly from the set of MNIST images.
+    Args:
+        test_size: Desired number of test images.
```
Added.

shrinuKushagra commented on examples/rcn.py:
```diff
@@ -0,0 +1,390 @@
+test_size = 20
+
+
+def fetch_mnist_dataset(test_size: int, seed: int = 5):
```
Added.

shrinuKushagra commented on examples/rcn.py:
```diff
@@ -0,0 +1,390 @@
+# %% [markdown]
+# # 1. Load the data
+# %%
+hps, vps = 12, 12
```
Added.

swaroopgj commented on examples/rcn.py:
```diff
@@ -0,0 +1,438 @@
+# %% [markdown]
+# RCN is a neuroscience based, data-efficient, interpretable object detection/vision system. It is a probabilistic graphical model (PGM) that trains in minutes and generalizes well (for certain traits). It can reason about, and is robust to, occlusion. It can also imagine and interpret scenes.
+#
+# Please refer to the [Science](https://www.cs.jhu.edu/~ayuille/JHUcourses/ProbabilisticModelsOfVisualCognition2020/Lec21/A%20generative%20vision%20model%20that%20trains%20with%20high%20data%20efficiency%20and%20breaks%20text-based%20CAPTCHAs.pdf) paper for more details. In this notebook, we implement a two-level RCN model on a small subset of the MNIST dataset.
```
:roll_eyes: 
Can you use this link please?
https://www.science.org/doi/10.1126/science.aag2612

@lazarox do we have open-access link to share?

swaroopgj replied:
I would also link to the earlier repo https://github.com/vicariousinc/science_rcn since you used that to train the model I assume.

shrinuKushagra replied:
Changed the link.

shrinuKushagra replied:
And added new link where the model is loaded.

swaroopgj commented on examples/rcn.py:
```diff
@@ -0,0 +1,438 @@
+    data["filters"],
+)
+
+hps, vps = 12, 12  # Horizontal and vertical pool sizes respectively for RCN models.
+num_orients = filters.shape[0]
+brightness_diff_threshold = 40.0
+max_perturb_radius = 22
```
I would re-group them and explain every variable defined here, with either an inline comment or a comment above it, like you did for pool sizes.
num_orients and bdt are related to preproc to run edge detection.
different from the core PGM model.

swaroopgj replied:
It also won't hurt to expand the pool size to indicate what it exactly is, is it half the window size, is it radius or full size?

shrinuKushagra replied:
Done.

shrinuKushagra commented on examples/rcn.py:
```diff
@@ -0,0 +1,438 @@
+
+
+# %%
+def fetch_mnist_dataset(test_size: int, seed: int = 5) -> Tuple[np.ndarray, np.ndarray]:
```
Seems that using `tuple` was enabled in python 3.9?
https://www.python.org/dev/peps/pep-0585/

I was trying to use the suggestion but keep getting `TypeError: 'type' object is not subscriptable`. The project is using python version of `3.7.5`, maybe thats why?

wlehrach replied:
Ahh - my bad. You need to add:
```
from __future__ import annotations
```
to the top of the file - it'll be default in future python version
You often want that for other reasons, as various other typing things will fail otherwise.

shrinuKushagra replied:
Fixed :D

shrinuKushagra commented on examples/rcn.py:
```diff
@@ -0,0 +1,438 @@
+    data["suppression_masks"],
+    data["filters"],
+)
+
```
Added detailed description of the parameters.
Moved `max_perturb_radius` to the relevant cell as it not central to the model and only used to save computation.

StannisZhou commented on examples/rcn.py:
```diff
@@ -0,0 +1,452 @@
+from jax import tree_util
+from joblib import Memory
+
+memory = Memory("./example_data/tmp")
```
Does this have to be here? If not maybe move it after all the imports

StannisZhou replied:
Seems like we are ignoring E402 in flake8 for some reason....Better to move it after all imports

shrinuKushagra replied:
Moved to after imports.

StannisZhou commented on examples/rcn.py:
```diff
@@ -0,0 +1,452 @@
+# %% [markdown]
+# We load the following variables (that have been pre-computed).
+# - train_set and train_labels - A sample of MNIST train dataset containing 100 train images and their labels.
+# - frcs and edges - Used to represent the learned rcn graphical models.
```
Link to the learning module in science RCN repo?

shrinuKushagra replied:
Added.

StannisZhou commented on examples/rcn.py:
```diff
@@ -0,0 +1,452 @@
+# ## 4.2 Make factors
+
+# %% [markdown]
+# ### 4.2.1 Pre-compute the valid configs for different perturb radii.
```
Radii -> radius for consistency?

StannisZhou commented on examples/rcn.py:
```diff
@@ -0,0 +1,452 @@
+
+
+max_perturb_radius = (
+    22  # The maximum perturb radii for which to pre-compute the valid configs.
```
radii -> radius

StannisZhou commented on examples/rcn.py:
```diff
@@ -0,0 +1,452 @@
+
+
+# %% [markdown]
+# ## 5.2 Run maximum a-posteriori probability (map) inference on all test images
```
I believe the word probability is redundant here

swaroopgj replied:
My earlier comment was intended for potential typo in usage of "map product inference". 
But you can just use MAP inference here without having to expand.

shrinuKushagra replied:
Changed to say MAP inference.

StannisZhou commented on pyproject.toml:
```diff
@@ -26,6 +26,7 @@ scipy = "^1.2.3"
 jupytext = "^1.11.3"
 tqdm = "^4.61.0"
 sphinx = {version = "^3", optional = true}
+scikit-learn = "^1.0.1"
```
Reminder to also add joblib as part of dependency

shrinuKushagra replied:
Added.

StannisZhou commented on examples/rcn.py:
```diff
@@ -0,0 +1,452 @@
+
+frc, edge, train_img = frcs[img_idx], edges[img_idx], train_set[img_idx]
+ax[0].imshow(train_img[pad : 200 - pad, pad : 200 - pad], cmap="gray")
+ax[0].set_title("Sample training image", fontsize=20)
```
Increase fontsize to 40

shrinuKushagra replied:
Done.

StannisZhou commented on examples/rcn.py:
```diff
@@ -0,0 +1,452 @@
+    model_img[r1, c1] = 0
+    model_img[r2, c2] = 0
+    ax[1].text(
+        (c1 + c2) // 2 - pad, (r1 + r2) // 2 - pad, str(w), color="green", fontsize=15
```
Increase fontsize to 25

shrinuKushagra replied:
Done.

StannisZhou commented on examples/rcn.py:
```diff
@@ -0,0 +1,452 @@
+
+ax[1].axis("off")
+ax[1].imshow(model_img[pad : 200 - pad, pad : 200 - pad], cmap="gray")
+ax[1].set_title("Corresponding learned rcn model", fontsize=20)
```
Increase fontsize to 40

shrinuKushagra replied:
Done.

StannisZhou reviewed (commented):
Left some minor comments. LGTM for the most part. Will take a final look after the comments are addressed. Thanks for patiently addressing all the comments!

StannisZhou reviewed (approved):
LGTM. Excited to have this example ready!
